### PR TITLE
Add oidc implicit work flow

### PR DESCRIPTION
**Pull request by mallow111**
## Description

Add OIDC implicit workflow which support reponse_type: token, id_token
## Motivation and Context

By default, oidc provider provides code exchange flow in oauth2-proxy, there are some cases that need to use implicit workflow, in that case, the response type will be token id_token rather than code, this PR is trying to implement oidc implicit work flow. It will help people who need to use oidc implicit work flow.

## How Has This Been Tested?

Yes, I have been tested it our in docker-desktop cluster and our working cluster.
Running steps:

$ make build
$  ./oauth2-proxy --config test.cfg
$   while the content of test.cfg is :
```
provider = "oidc"
redirect_url = "http://localhost:8087"
oidc_issuer_url = "http://blah/auth/realms/xyz"
upstreams = [
    "http://localhost:9091/"
]
email_domains = [
    "*"
]
client_id = "your client id"
client_secret="your secret"
pass_access_token = true
cookie_secret = "some cookie secret"
skip_provider_button = true
cookie_secure=false
http_address="127.0.0.1:8087"
insecure_oidc_allow_unverified_email=true
use_oidc_implicit_flow = true
insecure_oidc_skip_issuer_verification=true
```
$ in your browser open locahost:8087, you should be asked to login with identity user name and password, if authNed you will be redirect to the upstream service
## Checklist:

- [ ] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by JoelSpeed** [comment]
Hi @mallow111, thanks for raising this PR. Could you please expand for me on when the implicit flow is required rather than the code exchange flow? 

Could you also resolve the linting and codeclimate issues that have been highlighted

**Comment by mallow111** [comment]
@JoelSpeed I can definitely resolve the conflict,  I dont think implicit flow is better than code flow, code flow is definitely safer, but there are some cases that you cannot talk to the identity issuer backend, the only communication is through front end, in that case, people cannot use code flow, to make it safer, users need to add TLS on oauth2 proxy, and use http between oauth2 proxy and upstream. Let me know if this makes sense to you.

File: providers/oidc.go [hunk 1]
```diff
@@ -34,12 +41,29 @@ func NewOIDCProvider(p *ProviderData) *OIDCProvider {
 var _ Provider = (*OIDCProvider)(nil)
 
 // Redeem exchanges the OAuth2 authentication token for an ID token
-func (p *OIDCProvider) Redeem(ctx context.Context, redirectURL, code string) (s *sessions.SessionState, err error) {
+func (p *OIDCProvider) RedeemStandardFlow(ctx context.Context, redirectURL, code string) (s *sessions.SessionState, err error) {
```
Comment (mallow111) [inline providers/oidc.go:44]: this function is supposed to be exactly the same as Redeem function, however the CI check complains it has 6 return, limit is 4 return, so add additional getToken function to decouple the original Redeem function.

**Comment by mallow111** [comment]
@JoelSpeed  have fixed the CI check, please let me know if there is anything else that I can do, thanks!

File: providers/oidc.go [hunk 2]
```diff
@@ -48,27 +72,53 @@ func (p *OIDCProvider) Redeem(ctx context.Context, redirectURL, code string) (s
...
-	token, err := c.Exchange(ctx, code)
+	token, err = c.Exchange(ctx, code)
+	return
+}
+
+// use oidc implicit flow rather than stand flow, no code exchange needed
+func (p *OIDCProvider) RedeemImplicitFlow(redirectURL, code string) (s *sessions.SessionState, err error) {
+	// parse the code and get accessToken and idToken
+	split := strings.Split(code, ",")
+	accessToken, idToken := split[0], split[1]
```
Comment (thomashargrove) [inline providers/oidc.go:83]: Do you want to check the length of the array first?

File: providers/provider_data.go [hunk 3]
```diff
@@ -19,15 +26,28 @@ type ProviderData struct {
...
 }
 
+type TokenRelatedConfig struct {
+	AuthorizationEndpoint string `json:"authorization_endpoint"`
+	TokenEndpoint         string `json:"token_endpoint"`
+	JwksURI               string `json:"jwks_uri"`
+}
+
+var (
+	// ParsedConfig is the parsed config
```
Comment (thomashargrove) [inline providers/provider_data.go:47]: I dont think this comment adds much value :)

File: providers/provider_data.go [hunk 4]
```diff
@@ -44,3 +64,52 @@ func (p *ProviderData) GetClientSecret() (clientSecret string, err error) {
...
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get response body, err:%v", err)
+	}
+
+	err = json.Unmarshal(body, &ParsedConfig)
+	if err != nil {
+		panic(err)
```
Comment (thomashargrove) [inline providers/provider_data.go:94]: why panic?

File: providers/provider_data.go [hunk 5]
```diff
@@ -44,3 +64,52 @@ func (p *ProviderData) GetClientSecret() (clientSecret string, err error) {
...
+		AuthStyle: oauth2.AuthStyleAutoDetect,
+	}
+
+	keySet, err := jwk.Fetch(ParsedConfig.JwksURI)
+	if err != nil {
+		panic(err)
+	}
+
+	var signingKey rsa.PublicKey
+	err = keySet.Keys[0].Raw(&signingKey)
```
Comment (thomashargrove) [inline providers/provider_data.go:109]: check length of Keys[] first?

File: providers/provider_data.go (hunk 4 above)
Comment (thomashargrove) [reply providers/provider_data.go:94]: same with the next 2

**Comment by neuralsandwich** [comment]
It is currently against the IEFT OAuth security practices to use the implicit flow.

https://tools.ietf.org/html/draft-ietf-oauth-security-topics-15#section-2.1.2

**Comment by JoelSpeed** [comment]
Based on @neuralsandwich's https://github.com/oauth2-proxy/oauth2-proxy/pull/606#issuecomment-643778106 here, I am hesitant about progressing with this PR. 

Personally I also think supporting this extra flow adds a lot of extra  complexity for very little benefit (I don't think many users will want this pattern, though I'm willing to be proved wrong).

@mallow111 Does the workaround explanation that @neuralsandwich describes work for you at all?

**Comment by mallow111** [comment]
I will close the PR based on the discussion above.

**Pull request closed without merging**